### PR TITLE
Merge supports IQueryable<T> targets

### PR DESCRIPTION
**Pull request by viceroypenguin**
- [x] General CTE support with Merge statements
- [x] `.AsCte()` expression supported as Merge target
- [x] Inform users to call `.AsCte()` if they attempt to use a complex query as the merge target. 
- [x] Fix LinqSerializer for `With` field
- [x] `.AsCte()` expression supported as Merge source

### CODE REVIEW

**Comment by MaceWindu** [comment]
Which databases support CTE/Query as target?
I'm checking docs and see:
- sql sever: table/view
- db2 iSeries/zOs: table/view
- sap hana: table
- ifx: table/view
- sap ase: table/view
- oracle: table/view
- firebird: table/view

Looks like only db2 LUW declares CTE support in v11.5 and something called `fullselect`

**Comment by MaceWindu** [comment]
ok, there is stuff like that for sql server https://aboutsqlserver.com/2011/10/09/sunday-t-sql-tip-merge-into-cte-as-the-target/

What I don't understand is why it is not documented here https://docs.microsoft.com/en-us/sql/t-sql/statements/merge-transact-sql?view=sql-server-ver15

**Comment by viceroypenguin** [comment]
Yeah, I've been using CTE as the target for years. Huge advantage when working w/ a detail table that needs to be MERGEd for a single parent pkid. Use the CTE to restrict to the subset for the pkid, and then deletes in the MERGE only delete for that pkid and not the entire rest of the table.

File: Source/LinqToDB/Linq/Builder/MergeBuilder.Merge.cs [hunk 1]
```diff
@@ -11,20 +12,25 @@ internal class Merge : MethodCallBuilder
...
 
 			protected override IBuildContext BuildMethodCall(ExpressionBuilder builder, MethodCallExpression methodCall, BuildInfo buildInfo)
 			{
 				// Merge(ITable<TTarget> target, string hint)
 				var target = builder.BuildSequence(new BuildInfo(buildInfo, methodCall.Arguments[0], new SelectQuery()) { AssociationsAsSubQueries = true });
 
-				var targetTable = ((TableBuilder.TableContext)target).SqlTable;
-
-				var merge = new SqlMergeStatement(targetTable)
+				if (!(target is TableBuilder.TableContext tableContext))
```
Comment (sdanyliv) [inline Source/LinqToDB/Linq/Builder/MergeBuilder.Merge.cs:24]: I think you need helper function to throw this error. Also you have to check that tableContext.SelectQuery.IsSimple.

**Comment by viceroypenguin** [comment]
NB: Fixes #1166

**Comment by viceroypenguin** [comment]
Unit Test `MergeUsingCteWhere()` failed, it couldn't bind the predicate lambda properly. I haven't run it in a while though. I'll rebase and try again today - see if it's been fixed through the other things we've done.

**Comment by viceroypenguin** [comment]
@MaceWindu can you take a look at why `MergeUsingCteWhere()` isn't working? It is essentially the same as `MergeUsingCteJoin()`, but the join is handled differently. In `MergeUsingCteJoin()`, it's using the two separate key functions, where `MergeUsingCteWhere()` uses a predicate lambda. There's something in how it's looking for the CTE `.Id` column that's throwing it off.

File: Source/LinqToDB/Linq/Builder/TableBuilder.CteTableContext.cs [hunk 2]
```diff
@@ -133,6 +133,8 @@ public override SqlInfo[] ConvertToSql(Expression? expression, int level, Conver
 					ConvertToSql(null, 0, ConvertFlags.All);
 				}
 
+				expression = SequenceHelper.CorrectExpression(expression, this, queryContext);
```
Comment (MaceWindu) [inline Source/LinqToDB/Linq/Builder/TableBuilder.CteTableContext.cs:136]: @sdanyliv : fix location

File: Source/LinqToDB/Linq/Builder/MergeBuilder.MergeInto.cs [hunk 3]
```diff
@@ -11,20 +12,27 @@ internal class MergeInto : MethodCallBuilder
 			protected override bool CanBuildMethodCall(ExpressionBuilder builder, MethodCallExpression methodCall, BuildInfo buildInfo)
 			{
 				return methodCall.Method.IsGenericMethod
-					&& LinqExtensions.MergeIntoMethodInfo == methodCall.Method.GetGenericMethodDefinition();
+					&& (   LinqExtensions.MergeIntoMethodInfo1 == methodCall.Method.GetGenericMethodDefinition()
+						|| LinqExtensions.MergeIntoMethodInfo2 == methodCall.Method.GetGenericMethodDefinition());
```
Comment (sdanyliv) [inline Source/LinqToDB/Linq/Builder/MergeBuilder.MergeInto.cs:16]: Better to do that as in other Builders.

File: Source/LinqToDB/LinqExtensions.Merge.cs [hunk 4]
```diff
@@ -17,8 +17,10 @@ public static partial class LinqExtensions
...
+		internal static readonly MethodInfo MergeIntoMethodInfo1                      = MemberHelper.MethodOf(() => MergeInto<int, int>(null!, null!, null!))                            .GetGenericMethodDefinition();
+		internal static readonly MethodInfo MergeIntoMethodInfo2                      = MemberHelper.MethodOf(() => MergeInto<int, int>(null!, (IQueryable<int>)null!))                  .GetGenericMethodDefinition();
 		internal static readonly MethodInfo UsingMethodInfo1                          = MemberHelper.MethodOf(() => Using<int, int>(null!, (IQueryable<int>)null!))                      .GetGenericMethodDefinition();
 		internal static readonly MethodInfo UsingMethodInfo2                          = MemberHelper.MethodOf(() => Using<int, int>(null!, (IEnumerable<int>)null!))                     .GetGenericMethodDefinition();
 		internal static readonly MethodInfo UsingTargetMethodInfo                     = MemberHelper.MethodOf(() => UsingTarget<int>(null!))                                             .GetGenericMethodDefinition();
```
Comment (sdanyliv) [inline Source/LinqToDB/LinqExtensions.Merge.cs:26]: Move that to Methods.cs

File: Source/LinqToDB/SqlProvider/BasicSqlOptimizer.cs [hunk 5]
```diff
@@ -179,15 +179,39 @@ void FinalizeCte(SqlStatement statement)
 			{
 				var foundCte  = new Dictionary<CteClause, HashSet<CteClause>>();
 
-				select.SelectQuery.Visit(foundCte, static (foundCte, e) =>
-					{
-						if (e.ElementType == QueryElementType.SqlCteTable)
+				if (select is SqlMergeStatement merge)
```
Comment (sdanyliv) [inline Source/LinqToDB/SqlProvider/BasicSqlOptimizer.cs:182]: Why do we need here special processing?

**Comment by MaceWindu** [comment]
@viceroypenguin also would be nice to enable all merge providers for tests to check if any of them support CTE too.

**Comment by MaceWindu** [comment]
Made some tests with failing providers:
- SAP HANA: supports CTE in source query (CTE should be moved inside source clause as part of source query) but linq2db doesn't support CTE for SAP HANA right now
- Oracle: also supports cte if it is a part of source query. Also oracle doesn't like `_` as table alias
- DB2: version 11.5 we use on CI declares CTE support in target so it makes sense to test it (note that syntax is differ from sql server: https://www.ibm.com/docs/en/db2/11.5?topic=statements-merge). Also it makes sense to test if source CTE as part of select will work
- informix: CTE support added in v14 we use on CI, makes sense to test is as part of source query
- firebird: CTE works in source if moved to source query

so basically, for most of providers we should move source CTE to source query when we generate SQL and test CTE in target for DB2

File: Source/LinqToDB/SqlProvider/BasicSqlOptimizer.cs (hunk 5 above)
Comment (viceroypenguin) [reply Source/LinqToDB/SqlProvider/BasicSqlOptimizer.cs:182]: `SqlMergeStatement` inherits from `SqlStatementWithQueryBase` so that the `With` (CTE support) behavior will work properly. However, it does not use the `SelectQuery` property. It uses the `Target` and `Source` properties, which each have their own context and possible CTEs. This means it requires special support here to handle identifying CTEs in either the `Target` or `Source` branch. Possibly we could simplify to moving the target to the `SelectQuery` property, but that doesn't feel right.

File: Source/LinqToDB/Linq/Builder/MergeBuilder.MergeInto.cs (hunk 3 above)
Comment (viceroypenguin) [reply Source/LinqToDB/Linq/Builder/MergeBuilder.MergeInto.cs:16]: This code matches the other Merge builders (see Merge.InsertWhenNotMatched.cs, Merge.UpdateWhenMatched.cs, Merge.MergInto.cs, etc.).  I can change to doing it by name, but to match, I would need to do all of the merge builders similarly?

File: Source/LinqToDB/SqlProvider/BasicSqlOptimizer.cs (hunk 5 above)
Comment (sdanyliv) [reply Source/LinqToDB/SqlProvider/BasicSqlOptimizer.cs:182]: Why you have changed inheritance for `SqlMergeStatement` ?

File: Source/LinqToDB/Linq/Builder/MergeBuilder.MergeInto.cs (hunk 3 above)
Comment (viceroypenguin) [reply Source/LinqToDB/Linq/Builder/MergeBuilder.MergeInto.cs:16]: This currently matches the other Merge builders (Merge.InsertWhenNotMatched.cs, Merge.DeleteWhenMatched.cs, etc.). Would recommend that this is the preferred form personally. 

If you want, though, I can move all of the merge builders to the string versions? Or on a separate PR I can update the other builders in general to do by method comparison rather than string.

File: Source/LinqToDB/SqlProvider/BasicSqlOptimizer.cs (hunk 5 above)
Comment (viceroypenguin) [reply Source/LinqToDB/SqlProvider/BasicSqlOptimizer.cs:182]: so that the `With` (CTE support) behavior will work properly.

File: Source/LinqToDB/LinqExtensions.Merge.cs (hunk 4 above)
Comment (viceroypenguin) [reply Source/LinqToDB/LinqExtensions.Merge.cs:26]: Confirming: Do you want me to move this whole region to Methods.cs?

File: Source/LinqToDB/SqlProvider/BasicSqlOptimizer.cs (hunk 5 above)
Comment (sdanyliv) [reply Source/LinqToDB/SqlProvider/BasicSqlOptimizer.cs:182]: Yeah, under `Merge` section, something like that.

File: Tests/Linq/Update/MergeTests.IQueryableSource.cs [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+				AssertRow(InitialTargetData[0], result[0], null, null);
+				AssertRow(InitialTargetData[1], result[1], null, null);
+				AssertRow(InitialTargetData[2], result[2], null, 203);
+				AssertRow(InitialTargetData[3], result[3], null, null);
+				AssertRow(InitialSourceData[2], result[4], null, null);
+				AssertRow(InitialSourceData[3], result[5], null, 216);
+			}
+		}
+
+		[Test]
```
Comment (MaceWindu) [inline Tests/Linq/Update/MergeTests.IQueryableSource.cs:43]: disable tests for other providers with ActiveIssue(3015...)

File: Tests/Linq/Update/MergeTests.IQueryableSource.cs [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+		public void MergeFromCte([MergeDataContextSource] string context)
```
Comment (MaceWindu) [inline Tests/Linq/Update/MergeTests.IQueryableSource.cs:105]: disable tests for other providers with ActiveIssue(3015...)

File: Tests/Linq/Update/MergeTests.IQueryableSource.cs [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+				AssertRow(InitialTargetData[1], result[1], null, null);
+				AssertRow(InitialTargetData[2], result[2], null, 203);
+				AssertRow(InitialTargetData[3], result[3], null, null);
+				AssertRow(InitialSourceData[2], result[4], null, null);
+				AssertRow(InitialSourceData[3], result[5], null, 216);
+			}
+		}
+
+		[Test]
+		public void MergeUsingCteJoin([MergeDataContextSource] string context)
```
Comment (MaceWindu) [inline Tests/Linq/Update/MergeTests.IQueryableSource.cs:135]: disable tests for other providers with ActiveIssue(3015...)

File: Tests/Linq/Update/MergeTests.IQueryableSource.cs [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+				AssertRow(InitialTargetData[1], result[1], null, null);
+				AssertRow(InitialTargetData[2], result[2], null, 203);
+				AssertRow(InitialTargetData[3], result[3], null, null);
+				AssertRow(InitialSourceData[2], result[4], null, null);
+				AssertRow(InitialSourceData[3], result[5], null, 216);
+			}
+		}
+
+		[Test]
+		public void MergeUsingCteWhere([MergeDataContextSource] string context)
```
Comment (MaceWindu) [inline Tests/Linq/Update/MergeTests.IQueryableSource.cs:165]: disable tests for other providers with ActiveIssue(3015...)

**Review by MaceWindu — APPROVED** [review]

**Review by sdanyliv — APPROVED** [review]
